### DISCUSSION TITLE
LEDs not working with zmk

### DISCUSSION THREAD

**Discussion by dcolestock** [manna-harbour/miryoku#99] [Q&A] [answered]
I loaded a pretty basic miryoku_zmk onto my split corne with nice nano v2s using the workflow with just board, shield, and alpha settings onto my new board and haven't been able to get the LEDs to light up.  Everything else on all the layers is working as expected.  I tried hitting the external power toggle on the media layer in various combinations with the RGB toggle, but still get nothing.  Though, oddly, the RGB toggle types an 'f' (the alpha layer version of that key in dvorak).  None of the other keys in the media layer produce alphas.

This could be a hardware issue as this is the first thing I've loaded onto the board and haven't otherwise been able to confirm the LEDs actually work, but can you help me rule out user error or firmware issue? What are the normal steps to turning on the key LEDs after loading a fresh uf2 file?  Do they start on? Are they even supported by miryoku_zmk? Is the fact that my RGB toggle button produces an alpha unusual and potentially part of the problem?

**Comment by dcolestock** [comment]
Looks like I need to enable underglow: https://zmk.dev/docs/features/underglow#enabling-rgb-underglow

**Comment by manna-harbour** [comment, marked as answer]
In general, for keyboard configuration see https://github.com/manna-harbour/miryoku_zmk#kconfig-configuration.

In particular, RGB needs to be enabled for the keyboard first. See https://zmk.dev/docs/features/underglow#enabling-rgb-underglow. 

For corne, you can copy the default corne config file https://github.com/zmkfirmware/zmk/blob/main/app/boards/shields/corne/corne.conf below `config/` and uncomment the RGB related lines. 

For workflow builds you can also do that, or use the kconfig field, e.g. `CONFIG_ZMK_RGB_UNDERGLOW=y\nCONFIG_WS2812_STRIP=y`.

If you still get an alpha after RGB is enabled let me know.

**Comment by dcolestock** [comment]
I've managed to get underglow _almost_ working.   The primary split works great, but secondary split half won't respond to the external power switch or any of the RGB buttons except the RGB toggle.  The secondary split doesn't even respect underglow configuration settings such as `CONFIG_ZMK_RGB_UNDERGLOW_HUE_START=47` or `CONFIG_ZMK_RGB_UNDERGLOW_EFF_START=1`, so the secondary split really only does the solid red.  I assume this means that this is a feature still under development for split keyboards in the main ZMK library?

Activating this feature actually turned on RGBs one the back of my board that I didn't realize I have.  Not sure how to turn on the lights that are under each key, might have to play with the backlight feature or modifying the overlay.

### Steps to get RGB underglow _almost_ working in with ZMK

1. Find your [shield](https://github.com/zmkfirmware/zmk/blob/main/app/boards/shields) (ex. corne) and copy its contents into your `miryoku_zmk/config/` folder INCLUDING the boards sub directory, but EXCLUDING `<shield>.keymap` since you don't want to overwrite the miryoku keymap for your shield already in the `config/` directory.
2. Edit the `<shield>.conf` file and uncomment the lines
```
CONFIG_ZMK_RGB_UNDERGLOW=y
CONFIG_WS2812_STRIP=y
```
3. Because `config/boards/nice_nano.overlay` isn't my board, I just renamed this to my board `nice_nano_v2.overlay` hoping that it was similar enough to what I needed for the v2.

**Reply by manna-harbour** [reply]
I believe it should be fully working on splits now, if the board and shield definitions support it, but I haven't tried it personally.

You only need `<shield>.conf` or the equivalent lines in the workflow `kconfig`. You don't need to copy any other files if you aren't modifying them.

RGB with `nice_nano_v2` is an open issue. See https://github.com/zmkfirmware/zmk/issues/885. If the work-arounds mentiond there don't work it'd be best to ask on the ZMK discord.

**Comment by dcolestock** [comment]
You're right it looks like I didn't need to copy any files I wasn't modifying.  The only file I needed to copy was `boards/nice_nano.overlay` to `boards/nice_nano_v2.overlay` because it doesn't exist.  Doing this causes the second board to only be red like described above.  The link suggested the alternative of using `-b nice_nano@2`, but this caused the second board to not light up at all and the RGB toggle could only be turned on again.  I attempted using the https://github.com/malinges/zmk fork and got the same results with the second board not turning on.

**Reply by manna-harbour** [reply]
`-b nice_nano@2` is only for use with, and is required by, https://github.com/malinges/zmk/tree/nice_nano_revisions.

To see if it's a hardware issue, you could flash the central firmware to the right side. Alternatively, if the MCUs are socketed and you have pro-micros, you could test with QMK. Also, if the LEDs aren't configured at all, external power control should still be working.

Otherwise, I'd suggest asking on the ZMK or Nice Keyboards (nice!nano) discords.

You might also like to try with the default zmk-config. The problem is unlikely to be specific to Miryoku ZMK, but being able to refer to a more standard config for testing might make it easier for others to help.

**Comment by Just-Insane** [comment]
I recall reading somewhere that if BT or RGB commands are sent on the peripheral side, the commands will not be sent back, basically making the commands only function on the primary side. That's what I'm seeing anyways.

I'm going to test duplicating those functions on the primary side tomorrow and will update with what I find.

**Reply by manna-harbour** [reply]
There is a recently discovered issue. There's a fix in this temporary branch. When it's been tested I'll add it to the main branch.

https://github.com/manna-harbour/miryoku_zmk/tree/globalshiftfunctions-tmp#global-shift-functions

**Reply by Just-Insane** [reply]
I've tried this branch and it works to turn the display and LEDs off on both halves of the keyboard. Thanks for creating this project!

**Comment by dcolestock** [comment]
I updated my ZMK and miryoku repos to the latest version, and my LEDS are working on both sides now! Thanks!
